test/base-functor/Main.hs

```diff
@@ -0,0 +1,45 @@
+module Main (main) where
+
+import Control.Monad.Reader (runReader)
+import Covenant.Internal.Data
+  ( allComponentTypes,
+    hasRecursive,
+    mkBaseFunctor,
+  )
+import Covenant.Test
+  ( DataDeclFlavor (Poly1PolyThunks),
+    DataDeclSet (DataDeclSet),
+  )
+import Covenant.Type (datatypeName)
+import Data.Map.Strict qualified as M
+import Optics.Core (view)
+import Test.QuickCheck
+  ( Arbitrary (arbitrary),
+    Property,
+  )
+import Test.Tasty (defaultMain, testGroup)
+import Test.Tasty.QuickCheck (forAllBlind, testProperty)
+
+main :: IO ()
+main =
+  defaultMain . testGroup "BaseFunctors" $
+    [testProperty "All recursion is replaced in base functor transform" baseFunctorsContainNoRecursion]
+
+baseFunctorsContainNoRecursion :: Property
+baseFunctorsContainNoRecursion = forAllBlind (arbitrary @(DataDeclSet 'Poly1PolyThunks)) $ \(DataDeclSet decls) ->
```

Comment:
I would replace this with `forAllShrinkShow` and pass a function for formatting.